### PR TITLE
954 admiral strategy

### LINKED ISSUE #954: Refine admiral Strategy
We updated, extended, and removed many admiral functions in the last year and we had a lot of discussions on how flexible functions and parameters should be and on the level of nesting.

To reduce the number of updates and discussions the admiral strategy should be refined.

This helps also new contributors and developers of admiral extension packages.

Proposed roadmap:

- [x] Collect pain points examples (https://exbpbox.box.com/s/4a8zoc6zufi79n0iayxchfn2dtj6z475)
- [x] Discuss pain points identify root causes
- [x] Create a proposal of a refined admiral strategy
- [x] Discuss and agree on refined admiral strategy
- [x] Document the refined admiral strategy

### PR DESCRIPTION
**Pull request by rossfarrugia**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/) 
- [x] Updated relevant unit tests or have written new unit tests - See [Unit Test Guide](https://pharmaverse.github.io/admiral/articles/unit_test_guidance.html#writing-unit-tests-in-admiral-)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiral/articles/programming_strategy.html#deprecation-1)?
- [x] Update to all relevant roxygen headers and examples 
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md`
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue so that it closes after successful merging. 
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done!  Much love to your accomplishment!

closes #954

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
-* A toolbox of re-usable functions and utilities to create ADaM datasets using R scripts in a modular manner (an "opinionated" design strategy)
-* Pharmaceutical communities and companies are encouraged to contribute to `{admiral}` following the provided programming strategy and modular approach
-* All functions are documented, tested, include examples and are listed in the
-[Reference section](https://pharmaverse.github.io/admiral/reference/index.html)
-* Vignettes on how to create ADSL, BDS and OCCUR datasets, including example scripts
-* Vignettes for ADaM dataset specific functionality (i.e. dictionary coding, date imputation, SMQs ...)
+## Main Goal
 
-## Usage
+Provide data scientists with an open source, modularized toolbox with which to create ADaM datasets
```
Comment (bms63) [inline README.md:32]: Can we just say, Provide users or developers?

File: README.md [hunk 2]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+in R from. _As opposed to a “run 1 line and an ADaM appears” black-box solution or an attempt to
+automate ADaM._
 
-* Think of `{admiral}` as a toolbox of modular blocks (R functions) to create analysis derivations
-    * Each block has a stand alone purpose, i.e. each function provides a specific functionality
-    * Data Scientists can create their own blocks, i.e. create own R functions
-* Constructing ADaM dataset should become like building out of blocks that are based on `{admiral}` modular functions and user created modular functions
+One of the key aspects of `{admiral}` is its development by the users for the users.
+It gives an entry point for all data scientists to collaborate, co-create and contribute to a
```
Comment (bms63) [inline README.md:37]: Feel like our audience is statistical programmers, but I guess the data scientists are taking over the world!  Could we make it less specific to a job role?

File: README.md (hunk 1 above)
Comment (rossfarrugia) [reply README.md:32]: agree - happy to update these mentions all to just say "users"

File: README.md [hunk 3]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
-* One of our principle design decisions in creating `{admiral}` was to prioritize transparency and simplicity for our users - and not to let this ever become a "black-box" toolkit
-* We hope `{admiral}` offers a chance for users to be programmers - this is not a "run 1 line and an ADaM appears" solution or an attempt to automate ADaM
-* It is expected for companies to adopt `{admiral}` that a company-specific extension package would likely be needed (e.g. `{admiral.roche}` or `{admiral.gsk}`)
-    * Consider this an opportunity to influence company-specific standards, as `{admiral}` offers a bridge towards a more industry-aligned implementation of ADaM
-* From the core `{admiral}` package covering common functions and mostly safety templates, there will be further package extensions dedicated to certain disease area endpoints (e.g. `{admiral.onco}` or `{admiral.hiv}`)
+To set expectations: It is not our target that `{admiral}` will ever provide all possible solutions
+for all ADaM datasets outside of study specific needs. It depends on the user's collaboration
+and contribution to help grow over time to a toolkit library that is robust, easy to use and
```
Comment (bms63) [inline README.md:44]: Could the phrase toolkit library be linked out to the pharmaverse website?

File: README.md [hunk 4]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+our philosophy, which we try to adhere to across the `{admiral}` family of packages.
+There isn’t always a clear single, straight forward rule but there are guiding principles we adhere to.
+This manifesto helps show the considerations of our developers when making decisions.
+
+We have four priority design principles to achieve the main goal as follows:
+
+_**Usability**_
+
+All `{admiral}` functions should be easy to use. 
+* Our documentation is an absolute priority. Each function reference page should cover the purpose,
```
Comment (bms63) [inline README.md:80]: Remove the Our?  Just Documentation....

File: README.md [hunk 5]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+descriptions of each argument with permitted values, the expected input and output, with clear real-life
+examples - so that users don’t need to dig through code to find answers.
+* Vignettes then complement the functional documentation to help users see how best they can be applied to
```
Comment (bms63) [inline README.md:83]:
```suggestion
* Vignettes thatcomplement the functional documentation to help users see how best the functions can be applied to
```

File: README.md [hunk 6]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+and advise to wrap them into a function. However we would also like to avoid multi-layered functional nesting,
+so this needs to be considered carefully to keep the nesting of 3-4 functions an exception rather than the rule.
+
+_**Findability**_
+
+In a growing codebase, across a family of packages, we make every effort to make our functions easily findable.
+* We use consistent naming conventions across all our functions, and provide vignettes and ADaM templates that
+help users to get started and build familiarity. Each `{admiral}` family package website is searchable.
+* We avoid repetitive functions that will do similar tasks (as explained above with study day example).
+* Each package extension is kept focused on the specific scope, i.e. any common need that could be raised up
```
Comment (bms63) [inline README.md:110]:
```suggestion
* Each package extension is kept focused on the specific scope, e.g. features that can enabled across extension packages will be brought to the core {admiral} package. 
```

File: README.md [hunk 7]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+examples - so that users don’t need to dig through code to find answers.
+* Vignettes then complement the functional documentation to help users see how best they can be applied to
+achieve customary ADaM requirements.
+* Functions should be written and structured in a way that users are able to read, re-use or extend them
+for study specific purposes if needed (see Readability below).
+
+_**Simplicity**_
+
+Each function has a clear purpose.
+* We try not to ever design single functions that could achieve numerous very different derivations, as for
```
Comment (bms63) [inline README.md:91]: I want to have a deeper read of this section as I find it confusing but don't have time yet to make suggestions.  I get gist of it and agree with it just got confused...I might be a bit slow sorry

File: README.md [hunk 8]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+In a growing codebase, across a family of packages, we make every effort to make our functions easily findable.
+* We use consistent naming conventions across all our functions, and provide vignettes and ADaM templates that
+help users to get started and build familiarity. Each `{admiral}` family package website is searchable.
+* We avoid repetitive functions that will do similar tasks (as explained above with study day example).
+* Each package extension is kept focused on the specific scope, i.e. any common need that could be raised up
+to the core package will be.
+
+_**Readability**_
+
+We have a very prescriptive Programming Strategy guide that all our developers and contributors must follow,
```
Comment (bms63) [inline README.md:115]:
```suggestion
We have a very prescriptive Programming Strategy guide(link?) that all our developers and contributors must follow,
```

File: README.md [hunk 9]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+so that all our code has a high degree of consistency and readability.
+* We mandate use of tidyverse (e.g. dplyr) over similar functionality existing in base R.
```
Comment (bms63) [inline README.md:117]: Do we mandate this?  I thought we mandate tidyverse style guide?  I don't see this in the Programming Strategy Guide

File: README.md [hunk 10]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+* For the sections of any code that perform the actual derivation (e.g. besides assertions or basic utilities),
+we try to limit nesting of too many dependencies or functions.
+* Modularity is a focus - we don’t try to achieve too many steps in one, and all code has to be well commented.
+* We recognise that a user or a HA reviewer may have the wish to delve into the codebase (especially given this
```
Comment (bms63) [inline README.md:121]: what is HA?

File: vignettes/programming_strategy.Rmd [hunk 11]
```diff
@@ -65,16 +65,53 @@ case-sensitive. Thus they must be in lower case.
...
+Firstly, it is important to explain how we decide on the need for new derivation functions. 
+
+If a derivation rule or algorithm is common and highly similar across different variables/parameters
+(e.g. study day or duration) then we would provide a generic function that can be used to satisfy all
+the times this may be needed across different ADaMs. Similarly, if we feel that a certain derivation
+could be useful beyond a single purpose we also would provide a generic function (e.g. instead of a
+last known alive date function, we have an extreme date function where a user could find the last date
+from a selection, or for example the first).
+
+Otherwise if we feel that a derivation rule is a unique need or sufficiently complex to justify then we
```
Comment (bms63) [inline vignettes/programming_strategy.Rmd:79]:
```suggestion
Otherwise, if we feel that a derivation rule is a unique need or sufficiently complex to justify then we
```

File: README.md [hunk 12]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+
+## Admiral Manifesto
+For `{admiral}` and all extension packages, we prioritise providing our users with a **simple to adopt** toolkit
+that enables them to produce **readable** and **easily constructible** ADaM programs. The following explains
+our philosophy, which we try to adhere to across the `{admiral}` family of packages.
+There isn’t always a clear single, straight forward rule but there are guiding principles we adhere to.
+This manifesto helps show the considerations of our developers when making decisions.
+
+We have four priority design principles to achieve the main goal as follows:
+
```
Comment (bms63) [inline README.md:76]: Could the four design principles be put at the top?  Maybe bulleted after main goal as follows: 

* Usability
* Simplicity
* Findability 
* Readability

File: vignettes/programming_strategy.Rmd [hunk 13]
```diff
@@ -65,16 +65,53 @@ case-sensitive. Thus they must be in lower case.
...
+opt for a dedicated function for that specific variable/parameter (e.g. treatment-emergent flag for AEs).
+
+If certain variables are closely connected (e.g. an imputed date and the corresponding imputation flag)
+then a single function would provide both variables.
+
+If something needed for ADaM could be achieved simply via an existing tidyverse function, then we do not
+wrap this into an admiral function, as that would add an unnecessary extra layer for users.
+
+The following principles are key when designing a new function:
```
Comment (bms63) [inline vignettes/programming_strategy.Rmd:88]: Should these line up with the the four principles in the README?

* Usability
* Simplicity
* Findability 
* Readability

**Review by bms63 — CHANGES_REQUESTED** [review]
This looks awesome!  I'd like to have another pass on it, but it is very nice read and honestly a bit inspiring!

**Comment by rossfarrugia** [comment]
thanks @bms63 - will work through these now and then once the others leave any comments, you could then feel free to re-review anytime. agree it is inspiring - i really think we're building here a foundation of something that's going to influence the industry for the better!

File: README.md (hunk 3 above)
Comment (rossfarrugia) [reply README.md:44]: i think we meant here more that admiral itself is a library of code, rather than the pharmaverse which curates this and other such packages. i'll update here so clearer and will use the wording "asset library" to match the "al" of admiral

File: README.md [hunk 14]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+all listed in the [Reference section](https://pharmaverse.github.io/admiral/reference/index.html)
+* Vignettes on how to create ADSL, BDS and OCCUR datasets, including example scripts
+* Vignettes for ADaM dataset specific functionality (i.e. dictionary coding, date imputation,
+SMQs ...)
+
+## Types of Packages
+There will be 3 foreseeable types of `{admiral}` packages:
+* Core package - one package containing all core functions required to create ADaMs,
+usable by any company (i.e. general derivations, utility functions and checks for ADSL, OCCDS and BDS)
+* TA (Therapeutic Area) package extensions - one package per disease area with functions that are
```
[the remark this answers is not in the document]
Comment (rossfarrugia) [reply README.md:63]: agree

File: README.md (hunk 4 above)
Comment (rossfarrugia) [reply README.md:80]: agree

File: README.md (hunk 5 above)
Comment (rossfarrugia) [reply README.md:83]: agree

File: README.md (hunk 6 above)
Comment (rossfarrugia) [reply README.md:110]: i'll go with... e.g. features that are relevant across multiple extension packages will be moved to the core `{admiral}` package.

File: README.md (hunk 7 above)
Comment (rossfarrugia) [reply README.md:91]: please do Ben - this was probably one of the hardest section to put into words, BUT that being said if we can't describe our guidance around "simplicity" in a simple way then we've clearly missed the point! :D

File: README.md (hunk 8 above)
Comment (rossfarrugia) [reply README.md:115]: agree, will add link

File: README.md (hunk 9 above)
Comment (rossfarrugia) [reply README.md:117]: @thomas-neitmann i remember you asked for this. should i add to programming strategy too?

File: README.md (hunk 10 above)
Comment (rossfarrugia) [reply README.md:121]: Health Authority - included in full now

File: vignettes/programming_strategy.Rmd (hunk 11 above)
Comment (rossfarrugia) [reply vignettes/programming_strategy.Rmd:79]: agree

File: README.md (hunk 12 above)
Comment (rossfarrugia) [reply README.md:76]: Robin had something similar in his draft, but i actually felt when it was down on the homepage that it reads better without repeating these so i took that out. Let's see what others think, but i was conscious of not putting too many new sections on the homepage for fear of scaring people off from actually reading it all

File: README.md [hunk 15]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+
+All `{admiral}` functions should be easy to use. 
+* Documentation is an absolute priority. Each function reference page should cover the purpose,
+descriptions of each argument with permitted values, the expected input and output, with clear real-life
+examples - so that users don’t need to dig through code to find answers.
+* Vignettes that complement the functional documentation to help users see how best the functions can be
+applied to achieve customary ADaM requirements.
+* Functions should be written and structured in a way that users are able to read, re-use or extend them
+for study specific purposes if needed (see Readability below).
+
```
Comment (upadhyor) [inline README.md:87]: Should we add some verbiage around how we prioritize Usability over Simplicity - while our goal is to make the functions easy to use, we don't want to make it so simple that nobody will actually use it.  There may be cases where adding a parameter or two will lead to much greater use of the function by end users.  So we are always debating that right balance, but in the end Usability takes priority over Simplicity.

**Review by upadhyor — COMMENTED** [review]
Great food for thought, Ross.  Thank you for putting this together.

**Comment by rossfarrugia** [comment]
Thanks @upadhyor! More credit to @koegerr for the efforts before his holiday. I'll work through your comments tomorrow and then we can leave this PR open for now to see any further input from the team, before we merge ahead of the next release.

File: README.md (hunk 15 above)
Comment (rossfarrugia) [reply README.md:87]: @upadhyor any specific text you would add or change for this point? i'm finding it difficult to see where to update. my view is the 2 kind of go hand in hand. for me the sentence below that says _"if you as a user pick up a function with >10 different arguments then chances are it is going to be difficult to understand if this function could be applied for your specific need"_ gives the general idea we're getting at. basically if there is a common need for users then we can add things to admiral functions but not at the cost of making them so complex that nobody will understand them. also we have this line in the programming strategy which i think speaks to your point about not over-simplifying: _"If something needed for ADaM could be achieved simply via an existing tidyverse function, then we do not wrap this into an admiral function, as that would add an unnecessary extra layer for users."_

File: README.md (hunk 7 above)
Comment (rossfarrugia) [reply README.md:91]: i personally think we need this, as in the SAS macro based tools we've designed in the past for similar pieces they all are usable but often require intense trainings and people have to learn them over many years. often they would be fine for the very standard pieces but when you needed to go non-standard they would be difficult to debug and users would call them a "black-box". whereas for admiral to be successful industry-wide in the open source it has to be simple to pick up and very transparent. that's why so many of the admiral discussions often balance out the developer perspective (how to make the codebase most efficient and maintainable) vs the user perspective, and we often end up siding on what would the user prefer.

File: README.md (hunk 12 above)
Comment (koegerr) [reply README.md:76]: @bms63 I would be fine as is right now. Is it ok to close this from your end?

File: README.md (hunk 7 above)
Comment (koegerr) [reply README.md:91]: @bms63 just checking, would you like to change it? Any suggestion?

File: README.md (hunk 9 above)
Comment (koegerr) [reply README.md:117]: @bms63 do you think it needs a broader discussion? I think this is basically the direction we going, it just needs to be added to the programming strategy.

File: README.md [hunk 16]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+We will provide:
+* A toolbox of re-usable functions and utilities to create ADaM datasets using R scripts in a
+modular manner (an "opinionated" design strategy)
+* Pharmaceutical communities and companies are encouraged to contribute to `{admiral}` following
+the provided programming strategy and modular approach
+* Functions that are comprehensively documented and tested, including example calls - these are
+all listed in the [Reference section](https://pharmaverse.github.io/admiral/reference/index.html)
+* Vignettes on how to create ADSL, BDS and OCCUR datasets, including example scripts
+* Vignettes for ADaM dataset specific functionality (i.e. dictionary coding, date imputation,
```
Comment (bms63) [inline README.md:56]: Is it possible to link these out  to the vignettes?  If folks come to the github and scroll down to the REadme it would be a good chance to send them to the part of the website with user friendly vignettes.  The reference section makes my eyes hut

File: README.md [hunk 17]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+SMQs ...)
+
+## Types of Packages
+There will be 3 foreseeable types of `{admiral}` packages:
+* Core package - one package containing all core functions required to create ADaMs,
+usable by any company (i.e. general derivations, utility functions and checks for ADSL, OCCDS and BDS)
+* TA (Therapeutic Area) package extensions - one package per TA with functions that are
+specific to algorithms and requirements for that particular TA (e.g. `{admiralonco}`)
```
Comment (bms63) [inline README.md:64]: What about linking to admiralonco github?

File: README.md [hunk 18]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+* Company package extensions - specific needs and plug-ins for the company, such as access to metadata
+(e.g. `{admiralroche}` or `{admiralgsk}`)
+
+## Admiral Manifesto
+For `{admiral}` and all extension packages, we prioritise providing our users with a **simple to adopt** toolkit
+that enables them to produce **readable** and **easily constructible** ADaM programs. The following explains
+our philosophy, which we try to adhere to across the `{admiral}` family of packages.
+There isn’t always a clear single, straight forward rule but there are guiding principles we adhere to.
```
Comment (bms63) [inline README.md:72]:
```suggestion
There isn’t always a clear single, straightforward rule, but there are guiding principles we adhere to for admiral.
```

File: README.md [hunk 19]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+descriptions of each argument with permitted values, the expected input and output, with clear real-life
+examples - so that users don’t need to dig through code to find answers.
+* Vignettes that complement the functional documentation to help users see how best the functions can be
+applied to achieve customary ADaM requirements.
+* Functions should be written and structured in a way that users are able to read, re-use or extend them
+for study specific purposes if needed (see Readability below).
+
+_**Simplicity**_
+
+Each function has a clear purpose.
```
Comment (bms63) [inline README.md:90]:
```suggestion
All {admiral} functions have a clear purpose.
```

File: README.md [hunk 20]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+Importantly here though, functions cannot be too general and trying to fulfil multiple, complex purposes.
+Similarly, functions should not allow expressions as arguments that are used as code snippets in function calls.
+* We recommend to avoid copy and paste of complex computational algorithms or repetitive code like checks
+and advise to wrap them into a function. However we would also like to avoid multi-layered functional nesting,
+so this needs to be considered carefully to keep the nesting of 3-4 functions an exception rather than the rule.
+
+_**Findability**_
+
+In a growing codebase, across a family of packages, we make every effort to make our functions easily findable.
+* We use consistent naming conventions across all our functions, and provide vignettes and ADaM templates that
```
Comment (bms63) [inline README.md:107]:
```suggestion
* In a growing codebase, across a family of packages, we make every effort to make our functions easily findable.
* We use consistent naming conventions across all our functions, and provide vignettes and ADaM templates that
```

File: README.md [hunk 21]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+help users to get started and build familiarity. Each `{admiral}` family package website is searchable.
+* We avoid repetitive functions that will do similar tasks (as explained above with study day example).
+* Each package extension is kept focused on the specific scope, e.g. features that are relevant across multiple
+extension packages will be moved to the core `{admiral}` package.
+
+_**Readability**_
+
+We have a very prescriptive [Programming Strategy](https://pharmaverse.github.io/admiral/articles/programming_strategy.html)
```
Comment (bms63) [inline README.md:115]:
```suggestion
{All  {admiral} functions follow the [Programming Strategy](https://pharmaverse.github.io/admiral/articles/programming_strategy.html)
```

Comment (bms63) [reply README.md:115]: The other part can go as a separate bullet point

File: README.md [hunk 22]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+
+_**Usability**_
+
+All `{admiral}` functions should be easy to use. 
+* Documentation is an absolute priority. Each function reference page should cover the purpose,
+descriptions of each argument with permitted values, the expected input and output, with clear real-life
+examples - so that users don’t need to dig through code to find answers.
+* Vignettes that complement the functional documentation to help users see how best the functions can be
+applied to achieve customary ADaM requirements.
+* Functions should be written and structured in a way that users are able to read, re-use or extend them
```
Comment (bms63) [inline README.md:85]: I think this should be the first point

File: README.md [hunk 23]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+We have four priority design principles to achieve the main goal as follows:
+
+_**Usability**_
+
+All `{admiral}` functions should be easy to use. 
+* Documentation is an absolute priority. Each function reference page should cover the purpose,
+descriptions of each argument with permitted values, the expected input and output, with clear real-life
+examples - so that users don’t need to dig through code to find answers.
+* Vignettes that complement the functional documentation to help users see how best the functions can be
+applied to achieve customary ADaM requirements.
```
Comment (bms63) [inline README.md:84]:
```suggestion
applied to achieve ADaM requirements.
```

File: README.md (hunk 17 above)
Comment (rossfarrugia) [reply README.md:64]: i wouldn't personally just yet as the site would be the main thing useful to link to (the github repo is easily findable in pharmaverse org anyway) and the site is not quite ready yet for main stream attention. in future i can see us having something in the readme for a list of all such package extensions with each site link though

File: README.md [hunk 24]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+* Functions should be written and structured in a way that users are able to read, re-use or extend them
+for study specific purposes if needed (see Readability below).
+
+_**Simplicity**_
+
+Each function has a clear purpose.
+* We try not to ever design single functions that could achieve numerous very different derivations, as for
```
Comment (bms63) [inline README.md:91]:
```suggestion
* We try not to ever design single functions that could achieve numerous very different derivations. For
```

File: README.md [hunk 25]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+example if you as a user pick up a function with >10 different arguments then chances are it is going to be
+difficult to understand if this function could be applied for your specific need. The intention is that
+arguments/parameters can influence how the output of a function is calculated, but not change the purpose of
+the function.
+* On the other hand, we do try to combine similar tasks and algorithms into one function to reduce the amount
+of repetitive, similar functionality (e.g. one study day calculation rather than a function per variable).
+Importantly here though, functions cannot be too general and trying to fulfil multiple, complex purposes.
+Similarly, functions should not allow expressions as arguments that are used as code snippets in function calls.
```
Comment (bms63) [inline README.md:99]:
```suggestion
* We try to combine similar tasks and algorithms into one function to reduce the amount of repetitiveness and/or group together similar functionality (e.g. one study day calculation rather than a function per variable).
* We strive to design functions that are not too general and trying to fulfil multiple, complex purposes.
* Functions should not allow expressions as arguments that are used as code snippets in function calls.
```

Comment (bms63) [reply README.md:99]: ...this is hard.  I feel like declarative sentences are really important for a manifesto.  Some of these statements seem too wordy.

File: README.md (hunk 9 above)
Comment (bms63) [reply README.md:117]: I'm fine with that we should just state that somewhere.  As this is the first time I've seen this in writing.

File: README.md [hunk 26]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+* We avoid repetitive functions that will do similar tasks (as explained above with study day example).
+* Each package extension is kept focused on the specific scope, e.g. features that are relevant across multiple
+extension packages will be moved to the core `{admiral}` package.
+
+_**Readability**_
+
+We have a very prescriptive [Programming Strategy](https://pharmaverse.github.io/admiral/articles/programming_strategy.html)
+that all our developers and contributors must follow, so that all our code has a high degree of consistency and readability.
+* We mandate use of tidyverse (e.g. dplyr) over similar functionality existing in base R.
+* For the sections of any code that perform the actual derivation (e.g. besides assertions or basic utilities),
```
Comment (bms63) [inline README.md:118]:
```suggestion
* For sections of code that perform the actual derivations (e.g. besides assertions or basic utilities),
```

File: README.md [hunk 27]
```diff
@@ -11,47 +11,116 @@ ADaM in R Asset Library
...
+we try to limit nesting of too many dependencies or functions.
+* Modularity is a focus - we don’t try to achieve too many steps in one, and all code has to be well commented.
```
Comment (bms63) [inline README.md:120]:
```suggestion
* Modularity is a focus - we don’t try to achieve too many steps in one.
* All code has to be well commented.
```

File: vignettes/programming_strategy.Rmd (hunk 13 above)
Comment (bms63) [reply vignettes/programming_strategy.Rmd:88]: I guess I see them being the same.  The functions are the package and we have the 4 design principles that influence our practical approach to designing the functions.  Anyways, I am partial to short declarative sentences for designing stuff - as it is something I repeat to myself as I code things... e.g. "one proc away".  If you both are happy with how it reads, then I don't really have an issue.

File: vignettes/programming_strategy.Rmd [hunk 28]
```diff
@@ -65,16 +65,53 @@ case-sensitive. Thus they must be in lower case.
...
+steps is key for readability.
+
+* _**Avoid Copy and Paste**_ - If the same or very similar code is used multiple times, it should be put
+into a separate function. This improves readability and maintainability and makes unit testing easier.
+This should not be done for every simple programming step where tidyverse can be used. But rather for
+computational functions or data checks. However, also consider not to nest too many functions.
+
+* _**Checks**_ - Whenever a function fails, a meaningful error message must be provided with a clear
+reference to the input which caused the failure. This should avoid the users have to dig into detailed
+code if they only want to apply a function. This supports usability.
```
Comment (bms63) [inline vignettes/programming_strategy.Rmd:102]:
```suggestion
reference to the input which caused the failure. A users should not have to dig into detailed
code if they only want to apply a function.  A meaningful error message supports usability.
```

**Review by bms63 — APPROVED** [review]
Hi all, I made more comments and found some sentences that seemed incomplete.  Happy for you to incorporate my feedback or not!

File: README.md (hunk 9 above)
Comment (koegerr) [reply README.md:117]: I added it to the programming strategy too.

File: vignettes/programming_strategy.Rmd (hunk 13 above)
Comment (koegerr) [reply vignettes/programming_strategy.Rmd:89]: I see. Thanks Ben. We still have the whole "how to program a function, step by step" issue in the backlog. I think when we get to that we might need to adjust some wording as well :) I resolve this for now.

File: README.md (hunk 16 above)
Comment (koegerr) [reply README.md:56]: you mean to have them linked to the example scripts and vignettes directly from the readme?

File: README.md (hunk 25 above)
Comment (koegerr) [reply README.md:99]: I see your point. Let me go through it once more.

File: README.md (hunk 22 above)
Comment (koegerr) [reply README.md:85]: Do you mean the last bullet (85/86)?

File: README.md (hunk 25 above)
Comment (koegerr) [reply README.md:99]: @bms63 would that be better?

File: README.md (hunk 17 above)
Comment (bms63) [reply README.md:64]: makes sense thanks!

File: README.md (hunk 16 above)
Comment (bms63) [reply README.md:56]: yes

File: README.md (hunk 17 above)
Comment (koegerr) [reply README.md:64]: I close this for now.

**Comment by bms63** [comment]
@koegerr I spotted some inconsistencies in the formatting.  Getting nerd chills with how nice this is looking.  
image

File: README.md [hunk 29]
```diff
@@ -11,47 +11,118 @@ ADaM in R Asset Library
...
+* TA (Therapeutic Area) package extensions - one package per TA with functions that are
+specific to algorithms and requirements for that particular TA (e.g. `{admiralonco}`)
+* Company package extensions - specific needs and plug-ins for the company, such as access to metadata
+(e.g. `{admiralroche}` or `{admiralgsk}`)
+
+## Admiral Manifesto
+For `{admiral}` and all extension packages, we prioritise providing our users with a **simple to adopt** toolkit
+that enables them to produce **readable** and **easily constructible** ADaM programs. The following explains
+our philosophy, which we try to adhere to across the `{admiral}` family of packages.
+There isn’t always a clear single, straightforward rule, but there are guiding principles we adhere to for admiral.
```
Comment (bms63) [inline README.md:72]:
```suggestion
There isn’t always a clear single, straightforward rule, but there are guiding principles we adhere to for `{admiral}`.
```

**Review by thomas-neitmann — APPROVED** [review]
